### ISSUE TITLE
Multi-line find/replace project-wide (guidance)

### ISSUE THREAD

**Issue by thinkyhead** [atom/find-and-replace#824]
#398 seems to be stalled. I don't know the language used for this plugin or I would patch it myself. Nevertheless I'd like to give it a shot.

What files, functions, and lines should I focus on to fix this longstanding issue?

It would also be useful to expand the find/replace fields when the contents are multi-line. Where should I look to make that kind of change?

I'd also like to make it work more like Sublime Text (or have the option) so that it opens the files with matches, applies the replace, and leaves them open without saving. Where in the code would I implement such a hook?

Thanks for any guidance you can provide! I'm off to learn as much as I can about Coffeescript.

**Comment by 50Wliu** [comment]
This is not an actionable issue so I'm going to close this and suggest that you ask on [Slack](http://atom-slack.herokuapp.com/) or [Discuss, the official Atom and Electron message board](https://discuss.atom.io) instead.

However, to give you some pointers, I believe the core issue is that [Scandal](https://github.com/atom/scandal) does not support multiline searching as it breaks up the file into individual lines: https://github.com/atom/scandal/issues/5.

I thought expanding the mini-editors already worked by itself if you entered a newline.

The replacing appears to occur [here](https://github.com/atom/find-and-replace/blob/13df416278aaa83a6f2a2d806c1dd49d99b8c71c/lib/project/results-model.coffee#L145).

Again, I highly suggest that you post on either Slack or Discuss.  There's a bunch of active members on both that would be more than willing to help.

**Comment by thinkyhead** [comment]
Thanks, I will seek out the willing. Is Scandal an absolute requirement? I know that many Atom users are running with a full GNU shell environment, and it seems like it would be more-or-less "trivial" to make a thin wrapper for GNU utilities like `grep`, `sed`, `awk`, etc. to make a state-of-the-art find-and-replace.

Anyway, I will see what I can do through Slack, etc. I know there are a great many users who would love to have this solved.

**Comment by Ben3eeE** [comment]
> Just tested, and no. "Newline" initiates Find. The usual shortcuts (shift-Return, Alt-Return) don't insert a newline either.

What is supported is expanding for multiple lines if you paste multiple lines into the field. We have an open issue for adding newlines with the keyboard here https://github.com/atom/find-and-replace/issues/396.

**Comment by thinkyhead** [comment]
> paste

Ah, I will have to start using that, thanks!

**Comment by Ben3eeE** [comment]
This might be depending on the ui theme. I am using `one-dark`.

**Comment by thinkyhead** [comment]
Aha, good point! I will keep messing with Atom. I like it very much. Although writing plugins for it looks kind of scary. 😨

**Comment by Ben3eeE** [comment]
It's not scary just join the slack and ask any questions you might have. Did you check out the flight-manual? It has a great section on [hacking atom](http://flight-manual.atom.io/hacking-atom/) which describes how to write packages.

**Comment by thinkyhead** [comment]
I've been reading the flight manual, and it's very helpful, thanks! CoffeeScript looks cool. I'd've probably learned it sooner, but shortly after I got "CodeKit" (so I could get more into things like SASS and Less) I burned out on web development. But I have good Javascript chops so it's good to know I can implement packages in Javascript while I'm getting the hang of CoffeeScript. I'll definitely get on Slack and ask for help as I go along.

You were right about the theme. "Seti" doesn't expand the fields. Of course, soon I should be able to hack it to do whatever I want.

Thanks again!

— Scott out
